pygac/l1c_factory.py

```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env python
+
+# Copyright (c) 2014, 2019 Pygac Developers
+
+# Author(s):
+
+#   Carlos Horn <carlos.horn@external.eumetsat.int>
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Factory for l1c files.
+
+"""
+import datetime
+import logging
+
+from pygac.gac_klm import GACKLMReader
+from pygac.gac_pod import GACPODReader
+from pygac.lac_klm import LACKLMReader
+from pygac.lac_pod import LACPODReader
+from pygac.utils import file_opener
+
+
+LOG = logging.getLogger(__name__)
+
+
+class L1cBuilder(object):
+    def __init__(self, Reader):
+        self.Reader = Reader
+
+    def __call__(self, filename, start_line, end_line, fileobj=None):
+        tic = datetime.datetime.now()
+        LOG.info("Building file: %s", str(filename))
+        reader = self.Reader.fromfile(filename, fileobj=fileobj)
+        reader.save(start_line, end_line)
+        LOG.info("Processing took: %s", str(datetime.datetime.now() - tic))
+
+
+class L1cFactory(object):
+    """Factory for level 1c files."""
+    __builders = {
+        "gac_klm": L1cBuilder(GACKLMReader),
+        "gac_pod": L1cBuilder(GACPODReader),
+        "lac_klm": L1cBuilder(LACKLMReader),
+        "lac_pod": L1cBuilder(LACPODReader)
+    }
+
+    def __call__(self, filename, start_line, end_line, fileobj=None):
+        # check which builder to use
+        builder = self.get_builder(filename, fileobj=fileobj)
+        # process
+        builder(filename, start_line, end_line, fileobj=None)
+
+    @classmethod
+    def get_builder(cls, filename, fileobj=None):
+        try:
+            if fileobj is None:
+                open_file = file_opener(filename)
+            else:
+                open_file = file_opener(fileobj)
+            with open_file as fdes:
+                try:
+                    data = fdes.read(3).decode()
+                except UnicodeDecodeError:
+                    data = None
+        except IOError as err:
+            raise IOError('Failed to read GAC file: {0}'.format(err))
+        # TODO: discuss where this information belongs to.
+        # Maybe every reader needs a can_read class method
+        # which checks if this file belongs to the reader
+        # Then the factory loops over the reader classes to determine
+        # who can read it.
+        # At least, this desission does not belong into the run script!
```

Comment:
I absolutely agree! Having a mechanism to determine if a given reader can use the file is the cleanest solution.
Maybe some inspiration can be taken from this: https://github.com/pytroll/satpy/blob/master/satpy/readers/avhrr_l1b_gaclac.py#L103-L124
Also, I have stumbled upon cases where the file had been renamed, and the filename itself couldn't be trusted. However, the header of the files always contain the original name of the file, to we could parse this (using trollsift) to get more information what file it is, pod vs klm, gac vs lac.

Reply:
I will have a look and give it a try

Reply:
Thanks ! But that can be done in another PR, this one is already quite big as it is now :)